File: backend/app/schemas/package.py
"""Pydantic schemas for packages API."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import List, Literal, Optional
from pydantic import BaseModel, ConfigDict, Field, model_validator
from app.models.package import (
    PackageDefinitionStatus, EntitlementType, Shareability, PackageSaleStatus,
)


# ---------- Definition CRUD ----------

class PackageDefinitionItemCreate(BaseModel):
    service_id: str = Field(..., min_length=26, max_length=26)
    quantity: int = Field(default=1, ge=1)
    unit_price_paise: int = Field(..., ge=0)
    locked: bool = False
    display_order: int = 0
    max_redemptions: Optional[int] = Field(default=None, ge=1)


class DiscountInput(BaseModel):
    mode: Literal["pct", "flat", "final"]
    value: Decimal = Field(..., ge=0)


class PackageDefinitionCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    # v1 fields are optional so the v2 block builder (which sends `blocks` +
    # `final_price_paise` instead of items/entitlement_type) can reuse this
    # endpoint. v1 callers still send these and are validated as before.
    entitlement_type: EntitlementType = EntitlementType.COUNTED
    total_sessions: Optional[int] = Field(None, ge=1)
    shareability: Shareability = Shareability.OWNER_ONLY
    validity_days: int = Field(..., ge=1)
    auto_apply: bool = True
    cancellation_fee_pct: Decimal = Field(default=Decimal("20.00"), ge=0, le=100)
    items: List[PackageDefinitionItemCreate] = Field(default_factory=list)
    discount: Optional[DiscountInput] = None
    # Package Builder v2 — entitlement-block stack + builder-computed price.
    blocks: Optional[List[dict]] = None
    final_price_paise: Optional[int] = Field(None, ge=0)

    @model_validator(mode="after")
    def validate_entitlement_sessions(self) -> "PackageDefinitionCreate":
        if self.blocks is not None:
            return self  # v2 packages don't use the counted/sessions envelope
        if not self.items:
            raise ValueError("items must not be empty")
        if self.entitlement_type == EntitlementType.COUNTED and self.total_sessions is None:
            raise ValueError("total_sessions required when entitlement_type=counted")
        if self.entitlement_type == EntitlementType.UNLIMITED and self.total_sessions is not None:
            raise ValueError("total_sessions must be null when entitlement_type=unlimited")
        return self


class PackageDefinitionUpdate(PackageDefinitionCreate):
    pass


class PackageDefinitionItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    service_id: str
    service_name: Optional[str] = None
    quantity: int
    unit_price_paise: int
    locked: bool
    display_order: int
    max_redemptions: Optional[int] = None


class PackageDefinitionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: Optional[str]
    status: PackageDefinitionStatus
    entitlement_type: EntitlementType
    total_sessions: Optional[int]
    shareability: Shareability
    validity_days: int
    auto_apply: bool
    cancellation_fee_pct: Decimal
    items: List[PackageDefinitionItemResponse]
    # Items carry GROSS prices; discount + final_price_paise expose the
    # selling price so callers never re-implement distribution math.
    discount: Optional[DiscountInput] = None
    final_price_paise: int
    # v2 entitlement-block stack; null for v1 (items-based) packages.
    blocks: Optional[List[dict]] = None
    created_at: datetime
    updated_at: datetime


class PackageSaleItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    service_id: str
    service_name: Optional[str] = None
    quantity: int
    snapshot_unit_price_paise: int
    snapshot_gst_rate_pct: Decimal
    locked: bool
    max_redemptions: Optional[int] = None
    remaining: Optional[int] = None
    pool_exempt: bool = False


class PackageSaleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    bill_id: str
    package_definition_id: str
    package_definition_name: Optional[str] = None
    customer_id: str
    customer_name: Optional[str] = None
    selling_staff_id: Optional[str]
    sold_at: datetime
    expires_at: datetime
    entitlement_type_snapshot: EntitlementType
    shareability_snapshot: Shareability
    cancellation_fee_pct_snapshot: Decimal
    total_sessions_snapshot: Optional[int]
    sessions_remaining: Optional[int]
    status: PackageSaleStatus
    refunded_at: Optional[datetime]
    refund_bill_id: Optional[str]
    # v2 block stack as sold (null for v1); per-line consumption is on `items`.
    blocks_snapshot: Optional[List[dict]] = None
    created_at: datetime
    updated_at: datetime
    items: List[PackageSaleItemResponse]


class PackageSaleSummary(BaseModel):
    """Lightweight projection for eligibility rail."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    package_definition_name: Optional[str] = None
    entitlement_type_snapshot: EntitlementType
    sessions_remaining: Optional[int]
    total_sessions_snapshot: Optional[int]
    expires_at: datetime
    shareability_snapshot: Shareability
    customer_id: str
    customer_name: Optional[str] = None


# ---------- Eligibility ----------

class RedemptionEligibilityRequest(BaseModel):
    customer_id: str = Field(..., min_length=26, max_length=26)
    service_id: str = Field(..., min_length=26, max_length=26)


class EligiblePackageResponse(BaseModel):
    package_sale: PackageSaleSummary
    snapshot_price_paise: int
    # How many more times this service can be redeemed from this package (the
    # governing budget). The POS cart caps optimistic redemptions by this.
    remaining_uses: int = 0
    # Two-level budget for accurate cart allocation: this service's own cap, plus
    # a budget shared across services keyed by shared_budget_key (the package's
    # global pool, or a choice/pool block counter). null key = no shared budget.
    line_remaining: int = 0
    shared_budget_key: Optional[str] = None
    shared_remaining: int = 0


# ---------- Refund + extend ----------

class RefundRequest(BaseModel):
    payment_method: Literal["cash", "upi", "card", "pending_balance"]
    reason: str = Field(..., min_length=1)


class RefundBreakdown(BaseModel):
    kind: Literal["counted", "unlimited"]
    base_paise: int
    fee_paise: int
    refund_paise: int
    consumed_value_paise: int
    pct_remaining: Optional[Decimal] = None
    sessions_consumed: Optional[int] = None
    sessions_total: Optional[int] = None


class ExtendExpiryRequest(BaseModel):
    new_expires_at: datetime
    reason: str = Field(..., min_length=1)


class RefundResponse(BaseModel):
    credit_note_bill_id: str
    status: Literal["refunded"]

File: frontend/src/types/package.ts
// frontend/src/types/package.ts

export type PackageDefinitionStatus = "draft" | "published" | "archived";
export type EntitlementType = "counted" | "unlimited";
export type Shareability = "owner_only" | "shared";
export type PackageSaleStatus = "active" | "expired" | "refunded" | "exhausted";
export type DiscountMode = "pct" | "flat" | "final";

// ---------------------------------------------------------------------------
// Package Builder v2 — entitlement blocks
//
// A package is a stack of blocks. Each block kind carries its own entitlement
// semantics (replacing the old package-level `entitlement_type`). These are the
// BUILDER-SIDE draft shapes: money lives in paise (integers, per project rule),
// while small count fields (picks/sessions/daily_cap/quantity) are held as
// strings during editing — see the NumericCell draft-while-focused pattern.
// ---------------------------------------------------------------------------

export type BlockKind = "items" | "choice" | "unlimited" | "pool" | "credit";
export type ChooseAt = "purchase" | "visit";
export type CreditScope = "any" | "services" | "retail";

export interface ItemsRow {
  service_id: string;
  service_name: string; // display only
  quantity: string; // count, edited as string
  unit_price_paise: number;
}

export interface ChoiceRow {
  service_id: string;
  service_name: string;
  unit_price_paise: number;
}

export interface UnlimitedRow {
  service_id: string;
  service_name: string;
}

export interface PoolRow {
  service_id: string;
  service_name: string;
  unit_price_paise: number;
}

export type PackageBlock =
  | { id: string; kind: "items"; bonus: boolean; rows: ItemsRow[] }
  | {
      id: string;
      kind: "choice";
      bonus: boolean;
      picks: string;
      choose_at: ChooseAt;
      rows: ChoiceRow[];
    }
  | {
      id: string;
      kind: "unlimited";
      bonus: boolean;
      assigned_value_paise: number;
      daily_cap: string; // "" = no cap
      rows: UnlimitedRow[];
    }
  | { id: string; kind: "pool"; bonus: boolean; sessions: string; rows: PoolRow[] }
  | {
      id: string;
      kind: "credit";
      bonus: boolean;
      amount_paise: number;
      scope: CreditScope;
    };

export interface PackageDefinitionItem {
  id: string;
  service_id: string;
  service_name?: string;
  quantity: number;
  unit_price_paise: number;
  locked: boolean;
  display_order: number;
  max_redemptions: number | null;
}

export interface PackageDefinition {
  id: string;
  name: string;
  description?: string | null;
  status: PackageDefinitionStatus;
  entitlement_type: EntitlementType;
  total_sessions: number | null;
  shareability: Shareability;
  validity_days: number;
  auto_apply: boolean;
  cancellation_fee_pct: string; // Decimal serialized as string
  items: PackageDefinitionItem[];
  // Items carry gross prices; discount.value is paise for flat/final, % for pct
  discount: { mode: DiscountMode; value: string } | null;
  final_price_paise: number;
  // v2 entitlement-block stack; null for v1 (items-based) packages.
  blocks: PackageBlock[] | null;
  created_at: string;
  updated_at: string;
}

// Payload the v2 block builder sends to create/update a definition. Mirrors the
// loosened backend schema: blocks + builder-computed price, no items envelope.
export interface PackageDefinitionV2Payload {
  name: string;
  description?: string;
  validity_days: number;
  cancellation_fee_pct: string;
  shareability: Shareability;
  auto_apply: boolean;
  blocks: PackageBlock[];
  final_price_paise: number;
  discount?: { mode: DiscountMode; value: string };
}

export interface PackageDefinitionCreate {
  name: string;
  description?: string;
  entitlement_type: EntitlementType;
  total_sessions?: number;
  shareability: Shareability;
  validity_days: number;
  auto_apply: boolean;
  cancellation_fee_pct: string;
  items: Array<Omit<PackageDefinitionItem, "id" | "service_name">>;
  discount?: { mode: DiscountMode; value: string };
}

// PackageDefinitionCreate is also used for updates
export type PackageDefinitionUpdate = PackageDefinitionCreate;

export interface PackageSaleItem {
  id: string;
  service_id: string;
  service_name?: string;
  quantity: number;
  snapshot_unit_price_paise: number;
  snapshot_gst_rate_pct: string;
  locked: boolean;
  max_redemptions: number | null;
  remaining: number | null;
  pool_exempt: boolean;
}

export interface PackageSale {
  id: string;
  bill_id: string;
  package_definition_id: string;
  package_definition_name?: string;
  customer_id: string;
  customer_name?: string;
  selling_staff_id: string | null;
  sold_at: string;
  expires_at: string;
  entitlement_type_snapshot: EntitlementType;
  shareability_snapshot: Shareability;
  cancellation_fee_pct_snapshot: string;
  total_sessions_snapshot: number | null;
  sessions_remaining: number | null;
  status: PackageSaleStatus;
  refunded_at: string | null;
  refund_bill_id: string | null;
  blocks_snapshot: PackageBlock[] | null;
  items: PackageSaleItem[];
  created_at: string;
  updated_at: string;
}

export interface PackageSaleSummary {
  id: string;
  package_definition_name?: string | null;
  entitlement_type_snapshot: EntitlementType;
  sessions_remaining: number | null;
  total_sessions_snapshot: number | null;
  expires_at: string;
  shareability_snapshot: Shareability;
  customer_id: string;
  customer_name?: string | null;
}

export interface EligiblePackage {
  package_sale: PackageSaleSummary;
  snapshot_price_paise: number;
  /** How many more times this service can be redeemed from this package. */
  remaining_uses: number;
  /** This service's own per-line cap (independent of other services). */
  line_remaining: number;
  /** Key of a budget shared across services (global pool / block); null = none. */
  shared_budget_key: string | null;
  /** Remaining units in the shared budget. */
  shared_remaining: number;
}

export interface RefundBreakdown {
  kind: "counted" | "unlimited";
  base_paise: number;
  fee_paise: number;
  refund_paise: number;
  consumed_value_paise: number;
  pct_remaining?: string;
  sessions_consumed?: number;
  sessions_total?: number;
}

export interface RefundResponse {
  credit_note_bill_id: string;
  status: "refunded";
}
